### PR TITLE
feature(plus): enable api caching for all plus users

### PR DESCRIPTION
**Pull request by fiji-flo**
## Summary

Enable caching the API in IndexDB for all MDN Plus users.

### CODE REVIEW

**Review by LeoMcA — CHANGES_REQUESTED** [review]

File: client/pwa/src/service-worker.ts [hunk 1]
```diff
@@ -34,25 +38,25 @@ var unpacking = Promise.resolve();
...
-      const cache = await openCache();
-      const { files = {} } =
-        (await (await fetch("/asset-manifest.json")).json()) || {};
-      const assets = [...Object.values(files)].filter(
-        (asset) => !(asset as string).endsWith(".map")
-      );
-      await cache.addAll(assets as string[]);
-    })().then(() => self.skipWaiting())
+    SW_TYPE === SwType.ApiOnly
+      ? self.skipWaiting()
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply client/pwa/src/service-worker.ts:42]: We wanna do this automatically. And it's similar to consistently disabling the service worker. Another time :/

File: client/src/offline-settings/mdn-worker.tsx [hunk 2]
```diff
@@ -61,14 +61,14 @@ export class MDNWorker {
     this.controller()?.postMessage({ type: "update" });
   }
 
-  swName(onlineFirst: boolean | null | undefined = null) {
-    const onlineFirstSW = onlineFirst ?? this.settings.preferOnline ?? false;
+  swName(type: SwType | null | undefined = null) {
+    const onlineFirstSW = type ?? SwType.ApiOnly;
     return `/service-worker.js?preferOnline=${onlineFirstSW}`;
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply client/src/offline-settings/mdn-worker.tsx]: 👍

**Comment by fiji-flo** [comment]
Let me add dropping indexdb on active logout.

File: client/pwa/src/service-worker.ts [hunk 3]
```diff
@@ -25,34 +26,37 @@ const UPDATES_BASE_URL = `https://updates.${
   location.hostname === "localhost" ? "developer.allizom.org" : location.host
 }`;
 
+const SW_TYPE: SwType =
+  SwType[new URLSearchParams(location.search).get("type")] || SwType.ApiOnly;
```
Comment (LeoMcA) [inline client/pwa/src/service-worker.ts:30]: the left side of is returning a type (e.g. `ApiOnly`) and the right side is returning a number (e.g. `2`). I think `SwType.ApiOnly` needs to change into `SwType[SwType.ApiOnly]`

**Review by LeoMcA — CHANGES_REQUESTED** [review]
there's some enum type/value inconsistencies which I'm surprised typescript didn't detect and need fixing

also a super edge-case where the db isn't cleared if a user deletes their cookies, but not their indexed db - I'm not sure this is really reflective of real world conditions, though, as at least in firefox and chrome the non-devtools "clear site data" dialogue also clears the db

File: client/src/offline-settings/mdn-worker.tsx [hunk 4]
```diff
@@ -61,19 +68,28 @@ export class MDNWorker {
...
   }
 
-  async enableServiceWorker(onlineFirst: boolean | null | undefined = null) {
-    if ("serviceWorker" in navigator) {
-      await navigator.serviceWorker.register(this.swName(onlineFirst), {
+  registerMessageHandler() {
+    let self = this;
+    navigator.serviceWorker.addEventListener("message", (e) =>
+      self.messageHandler(e)
+    );
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply client/src/offline-settings/mdn-worker.tsx:79]: I thought that didn't work for some reason. But I think that was when I was calling this from somewhere else. Thanks.

File: client/src/offline-settings/mdn-worker.tsx [hunk 5]
```diff
@@ -61,19 +68,28 @@ export class MDNWorker {
...
+  }
```
[the remark this answers is not in the document]
Comment (fiji-flo) [reply client/src/offline-settings/mdn-worker.tsx:80]: 🤷‍♂️

File: client/pwa/src/service-worker.ts (hunk 3 above)
Comment (fiji-flo) [reply client/pwa/src/service-worker.ts:30]: Yes. I had string enums there and switched to numeric. Reverting that.

**Review by LeoMcA — APPROVED** [review]
lgtm, though a couple of small issues I observed in testing (which should probably be separate issues):

1. if a user loses their subscription status, but have already enabled mdn offline, they're stuck with their old settings with no way to change them

2. toggling offline on/off or prefer online on/off re-downloads the cached content, even if it's already there - at least in chrome

**Pull request merged**